app/grandchallenge/reader_studies/tasks.py

```diff
@@ -20,38 +18,33 @@
 )
 
 
-@transaction.atomic
-def add_score(obj, answer):
-    obj.calculate_score(answer)
-    obj.save(update_fields=["score"])
-
-
-@transaction.atomic
 def add_image(obj, image):
     obj.answer_image = image
     obj.save()
     image.assign_view_perm_to_creator()
     image.update_viewer_groups_permissions()
 
 
-@acks_late_micro_short_task
-def add_scores_for_display_set(*, instance_pk, ds_pk):
-    instance = Answer.objects.get(pk=instance_pk)
-    display_set = DisplaySet.objects.get(pk=ds_pk)
-    if instance.is_ground_truth:
-        for answer in Answer.objects.filter(
-            question=instance.question,
-            is_ground_truth=False,
-            display_set=display_set,
-        ):
-            add_score(answer, instance.answer)
-    else:
-        ground_truth = Answer.objects.filter(
-            question=instance.question,
-            is_ground_truth=True,
-            display_set=display_set,
-        ).get()
-        add_score(instance, ground_truth.answer)
+@acks_late_2xlarge_task
+@transaction.atomic
+def answers_from_ground_truth(*, reader_study_pk, target_user_pk):
+    reader_study = ReaderStudy.objects.get(pk=reader_study_pk)
+    target_user = get_user_model().objects.get(pk=target_user_pk)
+
+    all_answers = Answer.objects.filter(question__reader_study=reader_study)
+    if all_answers.filter(is_ground_truth=False, creator=target_user).exists():
+        raise ValueError("User already has answers")
+
+    for answer in all_answers.filter(is_ground_truth=True).all():
+        # Simplify permissions and create new answers
+        answer._state.adding = True
+        answer.id = None
+        answer.pk = None
+        answer.is_ground_truth = False
+        answer.creator = target_user
```

Comment:
This is potentially changing the creator. That should be kept the same and maybe just check that the target user is the same? That could be added to the filter above. Also note that the `.all()` is unnecessary.

Reply:
Like what was discussed in the last review, the owner shouldn't change. This task should not be concerned with permissions, those should be checked in the view and instead the task should only be responsible for carrying out the work it was asked to do.

Reply:
As discussed offline: overwriting the creator on the copy is reasonable.